### ISSUE TITLE
5.0.5 – memory leak still present

### ISSUE THREAD

**Issue by varyform** [phusion/passenger#1455]
5.0.5 (ent) still have memory leek. 1.5GB in ~8 hours.

![image](https://cloud.githubusercontent.com/assets/20039/6831612/196b2a7c-d32a-11e4-9c90-da99ef18701b.png)

Please let me know any additional info you need.

**Comment by FooBarWidget** [comment]
@OnixGH Can you reproduce this?

**Comment by FooBarWidget** [comment]
@varyform Thanks for reporting this problem. Could you help us by collecting some diagnostics information and answering a few questions?

1. How much traffic does your site get?
2. Can you show the output of `passenger-status`?
3. Can you show the output of `passenger-status --show=requests`?

**Comment by varyform** [comment]
@FooBarWidget 

~1.5 - 2k RPM

```
Version : 5.0.5
Date    : 2015-03-25 19:40:51 +0000
Instance: 4wAu9py1 (nginx/1.6.2 Phusion_Passenger/5.0.5)

----------- General information -----------
Max pool size : 20
Processes     : 17
Requests in top-level queue : 0

----------- Application groups -----------
/srv/apps/google/plus/public (production)#default:
  App root: /srv/apps/google/plus
  Requests in queue: 0
  * PID: 13047   Sessions: 1       Processed: 62504   Uptime: 4h 43m 27s
    CPU: 71%     Memory  : 632M    Last used: 1s ago
  * PID: 13059   Sessions: 1       Processed: 58725   Uptime: 4h 43m 26s
    CPU: 68%     Memory  : 658M    Last used: 0s ago
  * PID: 13072   Sessions: 1       Processed: 54256   Uptime: 4h 43m 26s
    CPU: 65%     Memory  : 639M    Last used: 0s ago
  * PID: 13085   Sessions: 1       Processed: 56395   Uptime: 4h 43m 25s
    CPU: 55%     Memory  : 677M    Last used: 0s ago
  * PID: 13117   Sessions: 1       Processed: 50881   Uptime: 4h 43m 24s
    CPU: 49%     Memory  : 572M    Last used: 3s ago
  * PID: 13231   Sessions: 1       Processed: 38252   Uptime: 4h 43m 24s
    CPU: 46%     Memory  : 571M    Last used: 0s ago
  * PID: 13344   Sessions: 1       Processed: 36107   Uptime: 4h 43m 23s
    CPU: 34%     Memory  : 575M    Last used: 0s ago
  * PID: 13363   Sessions: 0       Processed: 26984   Uptime: 4h 43m 22s
    CPU: 27%     Memory  : 545M    Last used: 1s ago
  * PID: 13378   Sessions: 0       Processed: 19501   Uptime: 4h 43m 21s
    CPU: 19%     Memory  : 495M    Last used: 1s ago
  * PID: 13392   Sessions: 0       Processed: 13598   Uptime: 4h 43m 21s
    CPU: 13%     Memory  : 538M    Last used: 7s ago
  * PID: 13406   Sessions: 0       Processed: 8303    Uptime: 4h 43m 20s
    CPU: 9%      Memory  : 505M    Last used: 6s ago
  * PID: 13422   Sessions: 0       Processed: 5477    Uptime: 4h 43m 19s
    CPU: 5%      Memory  : 505M    Last used: 6s ago
  * PID: 13512   Sessions: 0       Processed: 3125    Uptime: 4h 43m 18s
    CPU: 3%      Memory  : 440M    Last used: 16s ago
  * PID: 27820   Sessions: 0       Processed: 41      Uptime: 25m 0s
    CPU: 1%      Memory  : 328M    Last used: 1m 32s ago
  * PID: 28156   Sessions: 0       Processed: 24      Uptime: 24m 21s
    CPU: 0%      Memory  : 256M    Last used: 1m 53s ago
  * PID: 32229   Sessions: 0       Processed: 7       Uptime: 18m 17s
    CPU: 0%      Memory  : 216M    Last used: 3m 43s ago
  * PID: 8618    Sessions: 0       Processed: 0       Uptime: 3m 31s
    CPU: 0%      Memory  : 98M     Last used: 3m 31s ago
```

```
Version : 5.0.5
Date    : 2015-03-25 19:42:27 +0000
Instance: 4wAu9py1 (nginx/1.6.2 Phusion_Passenger/5.0.5)

{
   "thread1" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 11,
      "mbuf_pool" : {
         "active_blocks" : 145312,
         "active_memory" : {
            "bytes" : 74399744,
            "human_readable" : "71.0 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 4,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 2048,
            "human_readable" : "2.0 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 740195395,
      "total_clients_accepted" : 43004,
      "total_requests_begun" : 43004,
      "turbocaching" : {
         "fetches" : 1,
         "hit_ratio" : 0.0,
         "hits" : 0,
         "store_success_ratio" : 0.0,
         "store_successes" : 0,
         "stores" : 1
      }
   },
   "thread10" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 11,
      "mbuf_pool" : {
         "active_blocks" : 144751,
         "active_memory" : {
            "bytes" : 74112512,
            "human_readable" : "70.7 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 5,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 2560,
            "human_readable" : "2.5 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 787115827,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 0,
         "hit_ratio" : null,
         "hits" : 0,
         "store_success_ratio" : null,
         "store_successes" : 0,
         "stores" : 0
      }
   },
   "thread11" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 11,
      "mbuf_pool" : {
         "active_blocks" : 145354,
         "active_memory" : {
            "bytes" : 74421248,
            "human_readable" : "71.0 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 5,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 2560,
            "human_readable" : "2.5 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 912794113,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 1,
         "hit_ratio" : 0.0,
         "hits" : 0,
         "store_success_ratio" : 0.0,
         "store_successes" : 0,
         "stores" : 1
      }
   },
   "thread12" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 10,
      "mbuf_pool" : {
         "active_blocks" : 144859,
         "active_memory" : {
            "bytes" : 74167808,
            "human_readable" : "70.7 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 6,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 3072,
            "human_readable" : "3.0 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 767097875,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 0,
         "hit_ratio" : null,
         "hits" : 0,
         "store_success_ratio" : null,
         "store_successes" : 0,
         "stores" : 0
      }
   },
   "thread13" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 11,
      "mbuf_pool" : {
         "active_blocks" : 145151,
         "active_memory" : {
            "bytes" : 74317312,
            "human_readable" : "70.9 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 5,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 2560,
            "human_readable" : "2.5 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 797838442,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 0,
         "hit_ratio" : null,
         "hits" : 0,
         "store_success_ratio" : null,
         "store_successes" : 0,
         "stores" : 0
      }
   },
   "thread14" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 10,
      "mbuf_pool" : {
         "active_blocks" : 144910,
         "active_memory" : {
            "bytes" : 74193920,
            "human_readable" : "70.8 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 5,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 2560,
            "human_readable" : "2.5 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 832631580,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 0,
         "hit_ratio" : null,
         "hits" : 0,
         "store_success_ratio" : 0.0,
         "store_successes" : 0,
         "stores" : 0
      }
   },
   "thread15" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 11,
      "mbuf_pool" : {
         "active_blocks" : 144991,
         "active_memory" : {
            "bytes" : 74235392,
            "human_readable" : "70.8 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 4,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 2048,
            "human_readable" : "2.0 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 468360667,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 0,
         "hit_ratio" : null,
         "hits" : 0,
         "store_success_ratio" : null,
         "store_successes" : 0,
         "stores" : 0
      }
   },
   "thread16" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 11,
      "mbuf_pool" : {
         "active_blocks" : 145153,
         "active_memory" : {
            "bytes" : 74318336,
            "human_readable" : "70.9 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 209,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 107008,
            "human_readable" : "104.5 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 524458380,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 1,
         "hit_ratio" : 0.0,
         "hits" : 0,
         "store_success_ratio" : 0.0,
         "store_successes" : 0,
         "stores" : 1
      }
   },
   "thread17" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 10,
      "mbuf_pool" : {
         "active_blocks" : 145009,
         "active_memory" : {
            "bytes" : 74244608,
            "human_readable" : "70.8 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 6,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 3072,
            "human_readable" : "3.0 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 859269083,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 0,
         "hit_ratio" : null,
         "hits" : 0,
         "store_success_ratio" : null,
         "store_successes" : 0,
         "stores" : 0
      }
   },
   "thread18" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 10,
      "mbuf_pool" : {
         "active_blocks" : 144839,
         "active_memory" : {
            "bytes" : 74157568,
            "human_readable" : "70.7 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 7,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 3584,
            "human_readable" : "3.5 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 566540258,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 0,
         "hit_ratio" : null,
         "hits" : 0,
         "store_success_ratio" : null,
         "store_successes" : 0,
         "stores" : 0
      }
   },
   "thread19" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 10,
      "mbuf_pool" : {
         "active_blocks" : 144879,
         "active_memory" : {
            "bytes" : 74178048,
            "human_readable" : "70.7 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 5,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 2560,
            "human_readable" : "2.5 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 702965450,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 1,
         "hit_ratio" : 0.0,
         "hits" : 0,
         "store_success_ratio" : 0.0,
         "store_successes" : 0,
         "stores" : 1
      }
   },
   "thread2" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 12,
      "mbuf_pool" : {
         "active_blocks" : 144964,
         "active_memory" : {
            "bytes" : 74221568,
            "human_readable" : "70.8 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 4,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 2048,
            "human_readable" : "2.0 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 549227000,
      "total_clients_accepted" : 43004,
      "total_requests_begun" : 43004,
      "turbocaching" : {
         "fetches" : 2,
         "hit_ratio" : 0.0,
         "hits" : 0,
         "store_success_ratio" : 0.0,
         "store_successes" : 0,
         "stores" : 2
      }
   },
   "thread20" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 11,
      "mbuf_pool" : {
         "active_blocks" : 144957,
         "active_memory" : {
            "bytes" : 74217984,
            "human_readable" : "70.8 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 186,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 95232,
            "human_readable" : "93.0 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 661887533,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 1,
         "hit_ratio" : 0.0,
         "hits" : 0,
         "store_success_ratio" : 0.0,
         "store_successes" : 0,
         "stores" : 1
      }
   },
   "thread21" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 10,
      "mbuf_pool" : {
         "active_blocks" : 145088,
         "active_memory" : {
            "bytes" : 74285056,
            "human_readable" : "70.8 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 6,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 3072,
            "human_readable" : "3.0 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 768364696,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 1,
         "hit_ratio" : 0.0,
         "hits" : 0,
         "store_success_ratio" : 0.0,
         "store_successes" : 0,
         "stores" : 1
      }
   },
   "thread22" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 11,
      "mbuf_pool" : {
         "active_blocks" : 145042,
         "active_memory" : {
            "bytes" : 74261504,
            "human_readable" : "70.8 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 6,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 3072,
            "human_readable" : "3.0 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 928198141,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 0,
         "hit_ratio" : null,
         "hits" : 0,
         "store_success_ratio" : null,
         "store_successes" : 0,
         "stores" : 0
      }
   },
   "thread23" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 10,
      "mbuf_pool" : {
         "active_blocks" : 145291,
         "active_memory" : {
            "bytes" : 74388992,
            "human_readable" : "70.9 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 114,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 58368,
            "human_readable" : "57.0 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 913516689,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 1,
         "hit_ratio" : 0.0,
         "hits" : 0,
         "store_success_ratio" : 0.0,
         "store_successes" : 0,
         "stores" : 1
      }
   },
   "thread24" : {
      "active_client_count" : 1,
      "active_clients" : {
         "24-43003" : {
            "connected_at" : {
               "local" : "Wed Mar 25 19:42:27 2015",
               "relative" : "0s ago",
               "timestamp" : 1427312547.450968
            },
            "connection_state" : "ACTIVE",
            "current_request" : {
               "app_response_body_already_read" : {
                  "bytes" : 0,
                  "human_readable" : "0 bytes"
               },
               "app_response_body_fully_read" : true,
               "app_response_body_type" : "NO_BODY",
               "app_response_http_major" : 1,
               "app_response_http_minor" : 0,
               "app_response_http_state" : "PARSING_HEADERS",
               "app_response_want_keep_alive" : false,
               "app_sink_state" : {
                  "callback_in_progress" : false,
                  "initialized" : true,
                  "io_watcher_active" : false
               },
               "app_source_state" : {
                  "callback_in_progress" : false,
                  "initialized" : true,
                  "io_watcher_active" : true
               },
               "flags" : {
                  "dechunk_response" : true,
                  "https" : true,
                  "request_body_buffering" : false
               },
               "host" : "a.plus.google.com",
               "http_major" : 1,
               "http_minor" : 1,
               "http_state" : "COMPLETE",
               "method" : "GET",
               "path" : "/logout",
               "refcount" : 1,
               "request_body_already_read" : 0,
               "request_body_fully_read" : true,
               "request_body_type" : "NO_BODY",
               "response_begun" : false,
               "session" : {
                  "gupid" : "16afae1-MAwC6GRu6C",
                  "pid" : 13059
               },
               "session_checkout_try" : 1,
               "started_at" : {
                  "local" : "Wed Mar 25 19:42:27 2015",
                  "relative" : "0s ago",
                  "timestamp" : 1427312547.451179
               },
               "state" : "WAITING_FOR_APP_OUTPUT",
               "sticky_session" : false,
               "want_keep_alive" : false
            },
            "lingering_request_count" : 0,
            "name" : "24-43003",
            "number" : 43003,
            "output_channel_state" : {
               "bytes_buffered" : {
                  "bytes" : 0,
                  "human_readable" : "0 bytes"
               },
               "callback_in_progress" : false,
               "mode" : "IN_MEMORY_MODE",
               "nbuffers" : 0,
               "reader_state" : "RS_INACTIVE"
            },
            "refcount" : 2,
            "requests_begun" : 1
         }
      },
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 127,
      "free_request_count" : 10,
      "mbuf_pool" : {
         "active_blocks" : 144756,
         "active_memory" : {
            "bytes" : 74115072,
            "human_readable" : "70.7 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 2,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 1024,
            "human_readable" : "1.0 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 940537018,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 1,
         "hit_ratio" : 0.0,
         "hits" : 0,
         "store_success_ratio" : 0.0,
         "store_successes" : 0,
         "stores" : 1
      }
   },
   "thread3" : {
      "active_client_count" : 1,
      "active_clients" : {
         "3-43004" : {
            "connected_at" : {
               "local" : "Wed Mar 25 19:42:27 2015",
               "relative" : "0s ago",
               "timestamp" : 1427312547.532993
            },
            "connection_state" : "ACTIVE",
            "current_request" : {
               "app_response_body_already_read" : {
                  "bytes" : 0,
                  "human_readable" : "0 bytes"
               },
               "app_response_body_fully_read" : true,
               "app_response_body_type" : "NO_BODY",
               "app_response_http_major" : 1,
               "app_response_http_minor" : 0,
               "app_response_http_state" : "PARSING_HEADERS",
               "app_response_want_keep_alive" : false,
               "app_sink_state" : {
                  "callback_in_progress" : false,
                  "initialized" : true,
                  "io_watcher_active" : false
               },
               "app_source_state" : {
                  "callback_in_progress" : false,
                  "initialized" : true,
                  "io_watcher_active" : true
               },
               "content_length" : 142,
               "flags" : {
                  "dechunk_response" : true,
                  "https" : true,
                  "request_body_buffering" : false
               },
               "host" : "b.plus.google.com",
               "http_major" : 1,
               "http_minor" : 1,
               "http_state" : "PARSING_BODY",
               "method" : "POST",
               "path" : "/sessions",
               "refcount" : 1,
               "request_body_already_read" : 142,
               "request_body_fully_read" : true,
               "request_body_type" : "CONTENT_LENGTH",
               "response_begun" : false,
               "session" : {
                  "gupid" : "16afae1-VqInxuDESp",
                  "pid" : 13047
               },
               "session_checkout_try" : 1,
               "started_at" : {
                  "local" : "Wed Mar 25 19:42:27 2015",
                  "relative" : "0s ago",
                  "timestamp" : 1427312547.533120
               },
               "state" : "WAITING_FOR_APP_OUTPUT",
               "sticky_session" : false,
               "want_keep_alive" : false
            },
            "lingering_request_count" : 0,
            "name" : "3-43004",
            "number" : 43004,
            "output_channel_state" : {
               "bytes_buffered" : {
                  "bytes" : 0,
                  "human_readable" : "0 bytes"
               },
               "callback_in_progress" : false,
               "mode" : "IN_MEMORY_MODE",
               "nbuffers" : 0,
               "reader_state" : "RS_INACTIVE"
            },
            "refcount" : 2,
            "requests_begun" : 1
         }
      },
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 127,
      "free_request_count" : 10,
      "mbuf_pool" : {
         "active_blocks" : 144954,
         "active_memory" : {
            "bytes" : 74216448,
            "human_readable" : "70.8 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 1,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 512,
            "human_readable" : "512 bytes"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 538129886,
      "total_clients_accepted" : 43004,
      "total_requests_begun" : 43004,
      "turbocaching" : {
         "fetches" : 0,
         "hit_ratio" : null,
         "hits" : 0,
         "store_success_ratio" : 0.0,
         "store_successes" : 0,
         "stores" : 0
      }
   },
   "thread4" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 11,
      "mbuf_pool" : {
         "active_blocks" : 145350,
         "active_memory" : {
            "bytes" : 74419200,
            "human_readable" : "71.0 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 5,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 2560,
            "human_readable" : "2.5 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 688607697,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 0,
         "hit_ratio" : null,
         "hits" : 0,
         "store_success_ratio" : null,
         "store_successes" : 0,
         "stores" : 0
      }
   },
   "thread5" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 11,
      "mbuf_pool" : {
         "active_blocks" : 145121,
         "active_memory" : {
            "bytes" : 74301952,
            "human_readable" : "70.9 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 224,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 114688,
            "human_readable" : "112.0 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 840409398,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 1,
         "hit_ratio" : 0.0,
         "hits" : 0,
         "store_success_ratio" : 0.0,
         "store_successes" : 0,
         "stores" : 1
      }
   },
   "thread6" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 10,
      "mbuf_pool" : {
         "active_blocks" : 145002,
         "active_memory" : {
            "bytes" : 74241024,
            "human_readable" : "70.8 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 5,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 2560,
            "human_readable" : "2.5 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 890989599,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 1,
         "hit_ratio" : 0.0,
         "hits" : 0,
         "store_success_ratio" : 0.0,
         "store_successes" : 0,
         "stores" : 1
      }
   },
   "thread7" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 11,
      "mbuf_pool" : {
         "active_blocks" : 145267,
         "active_memory" : {
            "bytes" : 74376704,
            "human_readable" : "70.9 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 5,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 2560,
            "human_readable" : "2.5 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 800209305,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 0,
         "hit_ratio" : null,
         "hits" : 0,
         "store_success_ratio" : null,
         "store_successes" : 0,
         "stores" : 0
      }
   },
   "thread8" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 11,
      "mbuf_pool" : {
         "active_blocks" : 144984,
         "active_memory" : {
            "bytes" : 74231808,
            "human_readable" : "70.8 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 4,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 2048,
            "human_readable" : "2.0 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 1011456626,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 0,
         "hit_ratio" : null,
         "hits" : 0,
         "store_success_ratio" : null,
         "store_successes" : 0,
         "stores" : 0
      }
   },
   "thread9" : {
      "active_client_count" : 0,
      "active_clients" : {},
      "disconnected_client_count" : 0,
      "disconnected_clients" : {},
      "free_client_count" : 128,
      "free_request_count" : 11,
      "mbuf_pool" : {
         "active_blocks" : 145141,
         "active_memory" : {
            "bytes" : 74312192,
            "human_readable" : "70.9 MB"
         },
         "chunk_size" : 512,
         "free_blocks" : 57,
         "offset" : 448,
         "spare_memory" : {
            "bytes" : 29184,
            "human_readable" : "28.5 KB"
         }
      },
      "pid" : 31743,
      "server_state" : "ACTIVE",
      "total_bytes_consumed" : 874342789,
      "total_clients_accepted" : 43003,
      "total_requests_begun" : 43003,
      "turbocaching" : {
         "fetches" : 0,
         "hit_ratio" : null,
         "hits" : 0,
         "store_success_ratio" : null,
         "store_successes" : 0,
         "stores" : 0
      }
   },
   "threads" : 24
}
```

**Comment by dmitry** [comment]
@varyform what about file descriptors, are they left to be open?

**Comment by varyform** [comment]
@dmitry descriptors count is stable

**Comment by dmitry** [comment]
@varyform have you tried to look to the root`s descriptors (if you are running nginx under the root)?

**Comment by FooBarWidget** [comment]
@OnixGH This appears to be an mbuf leak. You can see in his JSON dump that the mbuf pool uses quite a lot of active memory.

**Comment by OnixGH** [comment]
@FooBarWidget yes, I had noticed that. This is a different issue than the file descriptor leaks we fixed previously, although @varyform please tell us if you do see anything increasing there (as root, as @dmitry mentioned).

**Comment by varyform** [comment]
@OnixGH, @dmitry 
`/proc/sys/fs/file-nr` is stable at around 3k
Anything else I should check?

**Comment by OnixGH** [comment]
Update: we've reproduced an mbuf memory leak of about 300 KB per minute per thread (max load). For an 18 thread setup that's over 300 MB per hour, so it looks like the order of leaking that's being reported here. Working on a fix ASAP.

**Comment by FooBarWidget** [comment]
I think I've been able to fix this issue with commit bec6f0a6c4f. OnixGH will perform more tests in the next few days to verify that the leak is indeed fixed.

Memory was leaked upon processing a request with multiple headers, upon processing a response with multiple headers, and upon processing a response with Set-Cookie headers. Every time such a request or response was processed, 512 bytes of memory was leaked due to improperly dereferencing relevant memory buffers.

**Comment by FooBarWidget** [comment]
Version 5.0.6 has been released: https://blog.phusion.nl/2015/03/31/passenger-5-0-6/
